zokrates_core/src/r1cs.rs

```diff
@@ -338,7 +336,13 @@ pub fn r1cs_program<T: Field>(
             FlatStatement::Directive(..) => continue
         }
     }
-    (variables, private_inputs_offset, a, b, c)
+    // Convert map back into list ordered by index
+    let mut variables_list = vec!["".to_string(); variables.len()];
```

Comment:
Couldn't we create this vector as we build the map? Then we wouldn't need to check that `variables.values()` is exactly 0..n.

Reply:
We certainly could. However that would mean for the entire r1cs program creation we would double the amount of memory needed to store the variable names even though it's only needed after the fact. Is there a way to reference the values, so they don't require more storage?

Reply:
How about this crate? https://docs.rs/bimap/0.1.5/bimap/
It uses two HashMaps of reference counting pointers internally so that the Strings would be allocated only once.